### PR TITLE
Pass -l:libfoo.a to linker directly.

### LINKED ISSUE #148: Cargo build say "library name must not be empty" while using statik(true)
run `cargo build` and get:
```
Compiling xxx-sys v0.1.0 (/xxx)
error: library name must not be empty

error: could not compile `xxx-sys` (lib) due to previous error
```

Run pkg-config manually:
```
# pkg-config --libs libdpdk --static
-Wl,--whole-archive -L/xxx -l:librte_bus_pci.a -l:librte_bus_vdev.a -l:librte_mempool_ring.a -l:librte_vhost.a -l:librte_security.a -l:librte_reorder.a -l:librte_dmadev.a -l:librte_power.a -l:librte_cryptodev.a -l:librte_compressdev.a -l:librte_timer.a -l:librte_hash.a -l:librte_cmdline.a -l:librte_pci.a -l:librte_ethdev.a -l:librte_meter.a -l:librte_net.a -l:librte_mbuf.a -l:librte_mempool.a -l:librte_rcu.a -l:librte_ring.a -l:librte_eal.a -l:librte_telemetry.a -l:librte_kvargs.a -Wl,--no-whole-archive -Wl,--export-dynamic -Wl,--as-needed -lrte_vhost -lrte_security -lrte_reorder -lrte_dmadev -lrte_power -lrte_cryptodev -lrte_compressdev -lrte_timer -lrte_hash -lrte_cmdline -lrte_pci -lrte_ethdev -lrte_meter -lrte_net -lrte_mbuf -lrte_mempool -lrte_rcu -lrte_ring -lrte_eal -lrte_telemetry -lrte_kvargs -pthread -lm -ldl
```

From $OUT_DIR/output:
```
cargo:rerun-if-env-changed=LIBDPDK_NO_PKG_CONFIG
cargo:rerun-if-env-changed=PKG_CONFIG_x86_64-unknown-linux-gnu
cargo:rerun-if-env-changed=PKG_CONFIG_x86_64_unknown_linux_gnu
cargo:rerun-if-env-changed=HOST_PKG_CONFIG
cargo:rerun-if-env-changed=PKG_CONFIG
cargo:rerun-if-env-changed=PKG_CONFIG_PATH_x86_64-unknown-linux-gnu
cargo:rerun-if-env-changed=PKG_CONFIG_PATH_x86_64_unknown_linux_gnu
cargo:rerun-if-env-changed=HOST_PKG_CONFIG_PATH
cargo:rerun-if-env-changed=PKG_CONFIG_PATH
cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR_x86_64-unknown-linux-gnu
cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR_x86_64_unknown_linux_gnu
cargo:rerun-if-env-changed=HOST_PKG_CONFIG_LIBDIR
cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR
cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_x86_64-unknown-linux-gnu
cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_x86_64_unknown_linux_gnu
cargo:rerun-if-env-changed=HOST_PKG_CONFIG_SYSROOT_DIR
cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR
cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR
cargo:rerun-if-env-changed=SYSROOT
cargo:rustc-link-search=native=/xxx
cargo:rustc-link-lib=:librte_bus_pci.a
cargo:rustc-link-lib=:librte_bus_vdev.a
cargo:rustc-link-lib=:librte_mempool_ring.a
cargo:rustc-link-lib=:librte_vhost.a
cargo:rustc-link-lib=:librte_security.a
cargo:rustc-link-lib=:librte_reorder.a
cargo:rustc-link-lib=:librte_dmadev.a
cargo:rustc-link-lib=:librte_power.a
cargo:rustc-link-lib=:librte_cryptodev.a
cargo:rustc-link-lib=:librte_compressdev.a
cargo:rustc-link-lib=:librte_timer.a
cargo:rustc-link-lib=:librte_hash.a
cargo:rustc-link-lib=:librte_cmdline.a
cargo:rustc-link-lib=:librte_pci.a
cargo:rustc-link-lib=:librte_ethdev.a
cargo:rustc-link-lib=:librte_meter.a
cargo:rustc-link-lib=:librte_net.a
cargo:rustc-link-lib=:librte_mbuf.a
cargo:rustc-link-lib=:librte_mempool.a
cargo:rustc-link-lib=:librte_rcu.a
cargo:rustc-link-lib=:librte_ring.a
cargo:rustc-link-lib=:librte_eal.a
cargo:rustc-link-lib=:librte_telemetry.a
cargo:rustc-link-lib=:librte_kvargs.a
cargo:rustc-link-lib=static=rte_vhost
cargo:rustc-link-lib=static=rte_security
cargo:rustc-link-lib=static=rte_reorder
cargo:rustc-link-lib=static=rte_dmadev
cargo:rustc-link-lib=static=rte_power
cargo:rustc-link-lib=static=rte_cryptodev
cargo:rustc-link-lib=static=rte_compressdev
cargo:rustc-link-lib=static=rte_timer
cargo:rustc-link-lib=static=rte_hash
cargo:rustc-link-lib=static=rte_cmdline
cargo:rustc-link-lib=static=rte_pci
cargo:rustc-link-lib=static=rte_ethdev
cargo:rustc-link-lib=static=rte_meter
cargo:rustc-link-lib=static=rte_net
cargo:rustc-link-lib=static=rte_mbuf
cargo:rustc-link-lib=static=rte_mempool
cargo:rustc-link-lib=static=rte_rcu
cargo:rustc-link-lib=static=rte_ring
cargo:rustc-link-lib=static=rte_eal
cargo:rustc-link-lib=static=rte_telemetry
cargo:rustc-link-lib=static=rte_kvargs
cargo:rustc-link-lib=m
cargo:rustc-link-lib=dl
cargo:rustc-link-arg=-Wl,--whole-archive
cargo:rustc-link-arg=-Wl,--no-whole-archive
cargo:rustc-link-arg=-Wl,--export-dynamic
cargo:rustc-link-arg=-Wl,--as-needed
cargo:rerun-if-env-changed=PKG_CONFIG_x86_64-unknown-linux-gnu
cargo:rerun-if-env-changed=PKG_CONFIG_x86_64_unknown_linux_gnu
cargo:rerun-if-env-changed=HOST_PKG_CONFIG
cargo:rerun-if-env-changed=PKG_CONFIG
cargo:rerun-if-env-changed=PKG_CONFIG_PATH_x86_64-unknown-linux-gnu
cargo:rerun-if-env-changed=PKG_CONFIG_PATH_x86_64_unknown_linux_gnu
cargo:rerun-if-env-changed=HOST_PKG_CONFIG_PATH
cargo:rerun-if-env-changed=PKG_CONFIG_PATH
cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR_x86_64-unknown-linux-gnu
cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR_x86_64_unknown_linux_gnu
cargo:rerun-if-env-changed=HOST_PKG_CONFIG_LIBDIR
cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR
cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_x86_64-unknown-linux-gnu
cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_x86_64_unknown_linux_gnu
cargo:rerun-if-env-changed=HOST_PKG_CONFIG_SYSROOT_DIR
cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR
cargo:rerun-if-changed=build.rs
cargo:rerun-if-changed=wrapper.h
```

For the error message "library name must not be empty", I found something in the rustc source:
https://github.com/rust-lang/rust/blob/71f71a5397c42fec01f5c1045c638d961fa9f7ca/compiler/rustc_session/src/config.rs#L2374
```rust
fn parse_libs(handler: &EarlyErrorHandler, matches: &getopts::Matches) -> Vec<NativeLib> {
    matches
        .opt_strs("l")
        .into_iter()
        .map(|s| {
            // Parse string of the form "[KIND[:MODIFIERS]=]lib[:new_name]",
            // where KIND is one of "dylib", "framework", "static", "link-arg" and
            // where MODIFIERS are a comma separated list of supported modifiers
            // (bundle, verbatim, whole-archive, as-needed). Each modifier is prefixed
            // with either + or - to indicate whether it is enabled or disabled.
            // The last value specified for a given modifier wins.
            let (name, kind, verbatim) = match s.split_once('=') {
                None => (s, NativeLibKind::Unspecified, None),
                Some((kind, name)) => {
                    let (kind, verbatim) = parse_native_lib_kind(handler, matches, kind);
                    (name.to_string(), kind, verbatim)
                }
            };

            let (name, new_name) = match name.split_once(':') {
                None => (name, None),
                Some((name, new_name)) => (name.to_string(), Some(new_name.to_owned())),
            };
            if name.is_empty() {
                handler.early_error("library name must not be empty");
            }
            NativeLib { name, new_name, kind, verbatim }
        })
        .collect()
}
```

And from rustc reference: https://doc.rust-lang.org/rustc/command-line-arguments.html#-l-link-the-generated-crate-to-a-native-library
```
[-l: link the generated crate to a native library](https://doc.rust-lang.org/rustc/command-line-arguments.html#-l-link-the-generated-crate-to-a-native-library)

Syntax: -l [KIND[:MODIFIERS]=]NAME[:RENAME].

This flag allows you to specify linking to a specific native library when building a crate.

The kind of library can optionally be specified with the form -l KIND=lib where KIND may be one of:

    dylib — A native dynamic library.
    static — A native static library (such as a .a archive).
    framework — A macOS framework.

If the kind is specified, then linking modifiers can be attached to it. Modifiers are specified as a comma-delimited string with each modifier prefixed with either a + or - to indicate that the modifier is enabled or disabled, respectively. Specifying multiple modifiers arguments in a single link attribute, or multiple identical modifiers in the same modifiers argument is not currently supported.
Example: -l static:+whole-archive=mylib.

The kind of library and the modifiers can also be specified in a [#[link] attribute](https://doc.rust-lang.org/reference/items/external-blocks.html#the-link-attribute). If the kind is not specified in the link attribute or on the command-line, it will link a dynamic library if available, otherwise it will use a static library. If the kind is specified on the command-line, it will override the kind specified in a link attribute.

The name used in a link attribute may be overridden using the form -l ATTR_NAME:LINK_NAME where ATTR_NAME is the name in the link attribute, and LINK_NAME is the name of the actual library that will be linked.
```

### PR DESCRIPTION
**Pull request by chenyuanrun**
Fix #148 .

### CODE REVIEW

**Comment by chenyuanrun** [comment]
Something need to be changed

**Comment by chenyuanrun** [comment]
> Something need to be changed

Done.

File: src/lib.rs [hunk 1]
```diff
@@ -726,7 +726,11 @@ impl Library {
                         continue;
                     }
 
-                    if statik && is_static_available(val, &system_roots, &dirs) {
+                    if val.starts_with(':') {
+                        // Pass this flag to linker directly.
+                        let meta = format!("cargo:rustc-link-arg={}{}", flag, val);
```
Comment (sdroege) [inline src/lib.rs:731]: This now would do `-l:foo.a` with a leading `:`. From what you wrote in the issue

> `Syntax: -l [KIND[:MODIFIERS]=]NAME[:RENAME].`

this doesn't seem valid, or am I missing something? You'd have an empty name then

Comment (chenyuanrun) [reply src/lib.rs:731]: cargo:rustc-link-arg will pass the flags to ld directly, this flag is not be handled by rustc.

Comment (chenyuanrun) [reply src/lib.rs:731]: -l:xxx link to a file named xxx, but there is no way to tell rustc to link to a file directly, if you pass -lxxx to rustc, rustc will try to link to libxxx.a instead , which is not what -l:xxx want

Comment (sdroege) [reply src/lib.rs:731]: Ah I see, that makes sense

**Review by sdroege — APPROVED** [review]

**Comment by chenyuanrun** [comment]
Thanks！

**Pull request merged**